File: docs/examples/sdk/agent/go/agent.go
// File-watching agent that classifies incoming documents with mlld.
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/fsnotify/fsnotify"
	mlld "github.com/mlld-lang/mlld/sdk/go"
)

func main() {
	scriptPath, _ := filepath.Abs(filepath.Join("..", "llm", "process.mld"))

	os.MkdirAll("inbox", 0o755)
	os.MkdirAll("done", 0o755)

	client := mlld.New()
	defer client.Close()

	watcher, err := fsnotify.NewWatcher()
	if err != nil {
		fmt.Fprintf(os.Stderr, "watcher error: %v\n", err)
		os.Exit(1)
	}
	defer watcher.Close()

	if err := watcher.Add("inbox"); err != nil {
		fmt.Fprintf(os.Stderr, "watch error: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("Watching inbox/ for new .md files. Drop a file in to classify it.")
	fmt.Println("Press Ctrl+C to stop.")
	fmt.Println()

	for {
		select {
		case event := <-watcher.Events:
			if event.Op&fsnotify.Create == 0 {
				continue
			}
			if filepath.Ext(event.Name) != ".md" {
				continue
			}

			// Brief delay to let file writes finish
			time.Sleep(200 * time.Millisecond)
			processFile(client, scriptPath, event.Name)

		case err := <-watcher.Errors:
			fmt.Fprintf(os.Stderr, "watcher error: %v\n", err)
		}
	}
}

func processFile(client *mlld.Client, scriptPath, filePath string) {
	name := filepath.Base(filePath)
	fmt.Printf("Processing %s...\n", name)

	content, err := os.ReadFile(filePath)
	if err != nil {
		fmt.Printf("  Error reading: %v\n", err)
		return
	}

	result, err := client.Execute(
		scriptPath,
		map[string]any{"content": string(content), "filename": name},
		&mlld.ExecuteOptions{Timeout: 60 * time.Second},
	)
	if err != nil {
		fmt.Printf("  Error: %v\n", err)
		return
	}

	for _, sw := range result.StateWrites {
		if sw.Path == "result" {
			out, _ := json.MarshalIndent(sw.Value, "", "  ")
			fmt.Printf("  -> %s\n", out)

			stem := name[:len(name)-len(filepath.Ext(name))]
			os.WriteFile(filepath.Join("done", stem+".result.json"), out, 0o644)
			break
		}
	}

	os.Rename(filePath, filepath.Join("done", name))
}

File: docs/examples/sdk/agent/rust/src/agent.rs
// File-watching agent that classifies incoming documents with mlld.

use notify::{Watcher, RecursiveMode, Event, EventKind};
use std::collections::HashMap;
use std::fs;
use std::path::Path;
use std::sync::mpsc;
use std::thread;
use std::time::Duration;

fn main() {
    let script_path = fs::canonicalize("../llm/process.mld")
        .expect("could not resolve ../llm/process.mld — run from the rust/ directory");
    let script = script_path.to_str().expect("invalid script path");

    fs::create_dir_all("inbox").ok();
    fs::create_dir_all("done").ok();

    let client = mlld::Client::new();

    let (tx, rx) = mpsc::channel::<Event>();
    let mut watcher = notify::recommended_watcher(move |res: Result<Event, _>| {
        if let Ok(event) = res {
            tx.send(event).ok();
        }
    })
    .expect("failed to create watcher");

    watcher
        .watch(Path::new("inbox"), RecursiveMode::NonRecursive)
        .expect("failed to watch inbox");

    println!("Watching inbox/ for new .md files. Drop a file in to classify it.");
    println!("Press Ctrl+C to stop.");
    println!();

    for event in rx {
        if !matches!(event.kind, EventKind::Create(_)) {
            continue;
        }

        for path in &event.paths {
            if path.extension().and_then(|e| e.to_str()) != Some("md") {
                continue;
            }

            // Brief delay to let file writes finish
            thread::sleep(Duration::from_millis(200));
            process_file(&client, script, path);
        }
    }
}

fn process_file(client: &mlld::Client, script: &str, path: &Path) {
    let name = path.file_name().unwrap().to_string_lossy();
    println!("Processing {name}...");

    let content = match fs::read_to_string(path) {
        Ok(c) => c,
        Err(e) => {
            println!("  Error reading: {e}");
            return;
        }
    };

    let mut payload = HashMap::new();
    payload.insert("content", content.as_str());
    payload.insert("filename", &name);

    let opts = mlld::ExecuteOptions::default().with_timeout(Duration::from_secs(60));

    match client.execute(script, Some(&payload), Some(opts)) {
        Ok(result) => {
            for sw in &result.state_writes {
                if sw.path == "result" {
                    let json = serde_json::to_string_pretty(&sw.value).unwrap_or_default();
                    println!("  -> {json}");

                    let stem = path.file_stem().unwrap().to_string_lossy();
                    fs::write(format!("done/{stem}.result.json"), &json).ok();
                    break;
                }
            }
            fs::rename(path, format!("done/{name}")).ok();
        }
        Err(e) => println!("  Error: {e}"),
    }
}
